### ISSUE TITLE
kOS and AGX

### ISSUE THREAD

**Issue by SirDiazo** [SirDiazo/AGExt#35]
@hvacengi In one of the kOS issue threads I saw you mention that you are doing an optimization pass and that the AGX code in kOS was on your list to look at.

(I closed the webpage and now can't find the thread again, it was one that came up on my watch list over the weekend.)

I currently don't have working test code to show, but I'm currently reworking AGX, (mainly for the new GUI system) and I am looking at changing how AGX interfaces with mods by using the new GameEvents system.

First, I'm not getting rid of the currently reflection system being used so feel free to keep using it.

However, the new GameEvents system looks to be simpler because supposedly if AGX creates an event and kOS subscribes to it, kOS can fire the event. That would be a lot cleaner then all the reflection code currently being used, at least in terms of amount of code required.

Now, as for actual questions:
1) Have you already done your pass of kOS's AGX code? I'm not trying to make a bunch of work for you here.
2) As someone who gets programming on a lot higher level then I do, does changing from using Reflection ourselves to using the GameEvents system (which uses Reflection anyway) make our lives better/easier?

I will also try to throw together some proof of concept code for you to actually see what I'm talking about tonight, wanted to get this issue opened so I don't forget about it.

**Comment by SirDiazo** [comment]
Bleh, not sure how far to follow this up, your comment about AGX was back at the start of December, the issue thread was closed over the weekend is why it showed up in my watched threads.

https://github.com/KSP-KOS/KOS/issues/1880

Still, would like your opinion on the new GameEvents system, but for actual use leave kOS on the reflection code it currently uses, I have no plans to remove it.

**Comment by hvacengi** [comment]
Well as far as changes I made to our AGX api, it was mostly limited to fixing an error that was being thrown when trying to get the type information (you can see it here: https://github.com/KSP-KOS/KOS/pull/1884/files#diff-ae0d073d81063dc5c63144abde6b0bb2 )

As for the new event system: I love that idea.  Reflection works OK, but is prone to breaking when the dependency assembly gets modified.  It's also an error trapping issue.  I personally have not yet taken the time to review the new GameEvents, but I think it's something that we at kOS definitely should be doing since it may simplify our interaction with some events.

That being said, there are some things that don't make a lot of sense to replicate as an event.  If we were to implement allowing users to edit AGX groups from within a kOS script, assigning an action to AG51 for example, that wouldn't make a lot of sense to be an event.  And anything that relies on a class within the dependency dll presents issues too.  You really can only pass objects that you know through events without running into reflection issues again anyways.

TLDR: I like the idea of supporting game events, I think it would make our AGX interaction cleaner, and I need to look into the system more.

**Comment by SirDiazo** [comment]
Alright.

It just got pushed back to tomorrow (i won't be home tonight now) but I'm going to try and prototype how game events would work for this and I'll probably actually move to a thread on the forums about it to get the biggest audience possible.

Are you active on the forums or do you want to comment on github?

**Comment by hvacengi** [comment]
I'm happy to chat on the forum, though once you know what you're doing I'll want to make an issue for kOS itself about it, otherwise it's too easy for me to loose track.  If you ping me I should see it within a couple days at worst.  I'll try to make sure I log in to the forums in the evenings to catch it.

**Comment by SirDiazo** [comment]
I'll certainly make a specific issue for kOS once I'm ready to move ahead, the forum thread is going to be a more general "this is what I want to do, does it make sense?" that I want feedback from the modders with more insight/access to Squad and the stock code-base.

**Comment by SirDiazo** [comment]
As this is looking to have larger implications then I expected, especially for kOS, I've opened a kOS issue here: https://github.com/KSP-KOS/KOS/issues/1939
